src/frontend/src/optimizer/rule/index_selection.rs

```diff
@@ -92,37 +100,175 @@ impl Rule for IndexSelectionRule {
                 let index_cost = index_table_scan_io_estimator.estimate(index_scan.predicate());
                 if index_cost.le(&min_cost) {
                     min_cost = index_cost;
-                    final_plan = index_scan;
+                    final_plan = index_scan.into();
                 }
             } else {
-                // TODO: non-covering index selection
+                // non-covering index selection
+                // only enable non-covering index selection when lookup join is enabled
+                let config = logical_scan.base.ctx.inner().session_ctx.config();
+                if config.get_batch_enable_lookup_join()
+                    && config.get_query_mode() == QueryMode::Local
+                {
+                    let (index_lookup, lookup_cost) =
+                        self.gen_index_lookup(logical_scan, index, p2s_mapping);
+                    if lookup_cost.le(&min_cost) {
+                        min_cost = lookup_cost;
+                        final_plan = index_lookup;
+                    }
+                }
             }
         }
 
+        // TODO: support merge index
+
         if min_cost == primary_cost {
             None
         } else {
-            Some(final_plan.into())
+            Some(final_plan)
+        }
+    }
+}
+
+struct Rewriter {
```

Comment:
Maybe we can name it specifically "IndexPredicateRewriter" etc.

Reply:
done